File: bench/cpp/ops/BenchStack.cpp
/*
 * SPDX-FileCopyrightText: Copyright (c) 2023-2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
 * SPDX-License-Identifier: Apache-2.0
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "../CppBenchUtils.hpp"
#include "ops/generated/BenchStackConfig.hpp"

#include <cvcuda/OpReformat.hpp>
#include <cvcuda/OpStack.hpp>

#include <nvbench/nvbench.cuh>

#include <vector>

template<typename BT>
inline void RunFakePlanarStackBench(nvbench::state &state, cvcuda::Stack &op, const int3 &shape, int ch)
{
    nvcv::TensorBatch         interSrc(nvcv::TensorBatch::CalcRequirements(shape.x));
    std::vector<nvcv::Tensor> planarSrcTensors;
    std::vector<nvcv::Tensor> interSrcTensors;
    planarSrcTensors.reserve(shape.x);
    interSrcTensors.reserve(shape.x);

    for (int i = 0; i < shape.x; ++i)
    {
        planarSrcTensors.emplace_back(nvcv::Tensor(
            {
                {ch, shape.y, shape.z},
                "CHW"
        },
            benchutils::GetDataType<BT>()));
        interSrcTensors.emplace_back(nvcv::Tensor(
            {
                {shape.y, shape.z, ch},
                "HWC"
        },
            benchutils::GetDataType<BT>()));
        benchutils::FillTensor<BT>(planarSrcTensors.back(), benchutils::CheckerboardValues<BT>());
        interSrc.pushBack(interSrcTensors.back());
    }

    nvcv::Tensor interDst(
        {
            {shape.x, shape.y, shape.z, ch},
            "NHWC"
    },
        benchutils::GetDataType<BT>());
    nvcv::Tensor dst(
        {
            {shape.x, ch, shape.y, shape.z},
            "NCHW"
    },
        benchutils::GetDataType<BT>());
    cvcuda::Reformat reformatOp;

    benchutils::warmup_and_exec(
        state, BENCH_STACK_WARMUP_ITERATIONS,
        [&op, &reformatOp, &planarSrcTensors, &interSrcTensors, &interSrc, &interDst, &dst](cudaStream_t s)
        {
            for (size_t i = 0; i < planarSrcTensors.size(); ++i)
            {
                reformatOp(s, planarSrcTensors[i], interSrcTensors[i]); // CHW -> HWC
            }
            op(s, interSrc, interDst);    // interleaved Stack
            reformatOp(s, interDst, dst); // NHWC -> NCHW
        });
}

template<typename T>
inline void stack(nvbench::state &state, nvbench::type_list<T>)
try
{
    int3                        shape     = benchutils::GetShape<3, int3>(state.get_string("shape"));
    auto                        layout    = benchutils::GetStringParam(state, "layout", "NHWC");
    const benchutils::InputKind inputKind = benchutils::GetInputKind(state.get_string("inputKind"));

    const bool isPlanar     = layout == "NCHW";
    const bool isFakePlanar = layout == "NCHW_FAKE";
    if (layout != "NHWC" && layout != "NCHW" && layout != "NCHW_FAKE")
    {
        state.skip("Stack benchmark supports only NHWC, NCHW, and NCHW_FAKE layouts");
        return;
    }
    if (isFakePlanar && inputKind != benchutils::InputKind::Tensor)
    {
        state.skip("Fake-planar (NCHW_FAKE) Stack benchmark is TensorBatch-only");
        return;
    }

    using BT = typename nvcv::cuda::BaseType<T>;

    int ch = nvcv::cuda::NumElements<T>;

    const long bytes = static_cast<long>(shape.x) * shape.y * shape.z * sizeof(T);
    state.add_global_memory_reads((isFakePlanar ? 3 : 1) * bytes);
    state.add_global_memory_writes((isFakePlanar ? 3 : 1) * bytes);

    cvcuda::Stack op;

    // clang-format off

    if (isFakePlanar)
    {
        RunFakePlanarStackBench<BT>(state, op, shape, ch);
    }
    else if (inputKind == benchutils::InputKind::Tensor)
    {
        nvcv::Tensor dst = isPlanar
                             ? nvcv::Tensor({{shape.x, ch, shape.y, shape.z}, "NCHW"}, benchutils::GetDataType<BT>())
                             : nvcv::Tensor({{shape.x, shape.y, shape.z, ch}, "NHWC"}, benchutils::GetDataType<BT>());
        nvcv::TensorBatch src(nvcv::TensorBatch::CalcRequirements(shape.x));

        for (int i = 0 ; i < shape.x; i++)
        {
            nvcv::Tensor srcIn = isPlanar
                                   ? nvcv::Tensor({{ch, shape.y, shape.z}, "CHW"}, benchutils::GetDataType<BT>())
                                   : nvcv::Tensor({{shape.y, shape.z, ch}, "HWC"}, benchutils::GetDataType<BT>());
            benchutils::FillTensor<BT>(srcIn, benchutils::CheckerboardValues<BT>());
            src.pushBack(srcIn);
        }

        benchutils::warmup_and_exec(state, BENCH_STACK_WARMUP_ITERATIONS,
            [&op, &src, &dst](cudaStream_t s) { op(s, src, dst); });
    }
    else
    {
        nvcv::Tensor dst = isPlanar
                             ? nvcv::Tensor({{shape.x, ch, shape.y, shape.z}, "NCHW"}, benchutils::GetDataType<BT>())
                             : nvcv::Tensor({{shape.x, shape.y, shape.z, ch}, "NHWC"}, benchutils::GetDataType<BT>());
        nvcv::ImageBatchVarShape src(shape.x);
        if (isPlanar && ch > 1)
        {
            benchutils::FillPlanarImageBatch<T>(src, long2{shape.z, shape.y}, long2{0, 0});
        }
        else
        {
            benchutils::FillImageBatch<T>(src, long2{shape.z, shape.y}, long2{0, 0},
                                          benchutils::CheckerboardValues<T>());
        }

        benchutils::warmup_and_exec(state, BENCH_STACK_WARMUP_ITERATIONS,
            [&op, &src, &dst](cudaStream_t s) { op(s, src, dst); });
    }

}
CVCUDA_BENCH_SKIP_ERRORS(state)

// clang-format on

// Use auto-generated type list from bench_params.json

NVBENCH_BENCH_TYPES(stack, NVBENCH_TYPE_AXES(BENCH_STACK_TYPES))
BENCH_STACK_AXES;

File: bench/python/ops/bench_stack.py
#!/usr/bin/env python3
# SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""CV-CUDA Stack operator benchmark - Python equivalent of BenchStack.cpp"""

import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent.parent))

import cvcuda  # noqa: E402
from python_bench_utils import (  # noqa: E402
    parse_shape,
    get_dtype,
    get_dtype_size,
    get_num_channels,
    get_format_from_dtype,
    get_input_kind,
    create_tensor,
    create_image_batch_varshape,
    create_stream_cache,
    run_benchmark,
)


def stack(state):
    """Stack operator benchmark matching C++ BenchStack.cpp"""

    shape = parse_shape(state.get_string("shape"))
    dtype_str = state.get_string("InOutDataType")
    dtype = get_dtype(dtype_str)
    input_kind = get_input_kind(state.get_string("inputKind"))
    try:
        layout = state.get_string("layout")
    except KeyError:
        layout = "NHWC"
    device_id = state.get_device()

    N, H, W = shape
    ch = get_num_channels(dtype_str)

    if layout not in ("NHWC", "NCHW", "NCHW_FAKE"):
        state.skip("Stack benchmark supports only NHWC, NCHW, and NCHW_FAKE layouts")
        return None
    is_planar = layout == "NCHW"
    is_fake_planar = layout == "NCHW_FAKE"
    if is_fake_planar and input_kind != "Tensor":
        state.skip("Fake-planar (NCHW_FAKE) Stack benchmark is TensorBatch-only")
        return None

    dtype_size = get_dtype_size(dtype_str)
    bytes_ = N * H * W * dtype_size
    state.add_global_memory_reads((3 if is_fake_planar else 1) * bytes_)
    state.add_global_memory_writes((3 if is_fake_planar else 1) * bytes_)

    get_stream = create_stream_cache()

    if is_fake_planar:
        planar_src_tensors = []
        inter_src_tensors = []
        inter_src = cvcuda.TensorBatch(N)
        for _ in range(N):
            planar_src = create_tensor(
                (ch, H, W), dtype, device_id, layout="CHW", fill_mode="checkerboard"
            )
            interleaved_src = create_tensor(
                (H, W, ch), dtype, device_id, layout="HWC", fill_mode=0
            )
            planar_src_tensors.append(planar_src)
            inter_src_tensors.append(interleaved_src)
            inter_src.pushback(interleaved_src)

        inter_dst = create_tensor(
            (N, H, W, ch), dtype, device_id, layout="NHWC", fill_mode=0
        )
        dst = create_tensor((N, ch, H, W), dtype, device_id, layout="NCHW", fill_mode=0)

        def run_fake(launch):
            stream = get_stream(launch)
            for planar_src, interleaved_src in zip(
                planar_src_tensors, inter_src_tensors, strict=True
            ):
                cvcuda.reformat_into(interleaved_src, planar_src, stream=stream)
            cvcuda.stack_into(inter_dst, inter_src, stream=stream)
            cvcuda.reformat_into(dst, inter_dst, stream=stream)

        return run_fake

    dst_shape = (N, ch, H, W) if is_planar else (N, H, W, ch)
    dst = create_tensor(dst_shape, dtype, device_id, layout=layout, fill_mode=0)

    if input_kind == "Tensor":
        src = cvcuda.TensorBatch(N)
        src_layout = "CHW" if is_planar else "HWC"
        src_shape = (ch, H, W) if is_planar else (H, W, ch)
        for _ in range(N):
            t = create_tensor(
                src_shape, dtype, device_id, layout=src_layout, fill_mode="checkerboard"
            )
            src.pushback(t)
    else:
        img_format = get_format_from_dtype(dtype_str, ch, planar=is_planar)
        src = create_image_batch_varshape(
            (N, H, W, ch),
            0,
            img_format,
            dtype,
            device_id,
            fill_mode="checkerboard",
        )

    def run(launch):
        stream = get_stream(launch)
        cvcuda.stack_into(dst, src, stream=stream)

    return run


if __name__ == "__main__":
    run_benchmark("stack", stack)
